source/Nabla/COpenGLDriver.cpp

```diff
@@ -252,173 +234,46 @@ bool COpenGLDriver::initDriver(CIrrDeviceWin32* device, EGLDisplay eglDisplay)
-
+        bool ocl_init = false;
 
-// -----------------------------------------------------------------------
-// LINUX CONSTRUCTOR
-// -----------------------------------------------------------------------
-#ifdef _NBL_COMPILE_WITH_X11_DEVICE_
-//! Linux constructor and init code
-COpenGLDriver::COpenGLDriver(const SIrrlichtCreationParameters& params,
-		io::IFileSystem* io, CIrrDeviceLinux* device, const asset::IGLSLCompiler* glslcomp)
-		: CNullDriver(device, io, Params), COpenGLExtensionHandler(),
-			runningInRenderDoc(false), ColorFormat(asset::EF_R8G8B8_UNORM),
-			X11Device(device), DeviceType(EIDT_X11), AuxContexts(0), GLSLCompiler(glslcomp)
-{
-	#ifdef _NBL_DEBUG
-	setDebugName("COpenGLDriver");
-	#endif
-}
-
-
-bool COpenGLDriver::changeRenderContext(const SExposedVideoData& videoData)
-{
-	if (videoData.OpenGLLinux.X11Window)
-	{
-		if (videoData.OpenGLLinux.X11Display && videoData.OpenGLLinux.X11Context)
-		{
-			if (!glXMakeCurrent((Display*)videoData.OpenGLLinux.X11Display, videoData.OpenGLLinux.X11Window, (GLXContext)videoData.OpenGLLinux.X11Context))
-			{
-				os::Printer::log("Render Context switch failed.");
-				return false;
-			}
-			else
-			{
-				Drawable = videoData.OpenGLLinux.X11Window;
-				X11Display = (Display*)videoData.OpenGLLinux.X11Display;
-			}
-		}
-		else
-		{
-			// in case we only got a window ID, try with the existing values for display and context
-			if (!glXMakeCurrent((Display*)ExposedData.OpenGLLinux.X11Display, videoData.OpenGLLinux.X11Window, (GLXContext)ExposedData.OpenGLLinux.X11Context))
-			{
-				os::Printer::log("Render Context switch failed.");
-				return false;
-			}
-			else
-			{
-				Drawable = videoData.OpenGLLinux.X11Window;
-				X11Display = (Display*)ExposedData.OpenGLLinux.X11Display;
-			}
-		}
-	}
-	// set back to main context
-	else if (X11Display != ExposedData.OpenGLLinux.X11Display)
-	{
-		if (!glXMakeCurrent((Display*)ExposedData.OpenGLLinux.X11Display, ExposedData.OpenGLLinux.X11Window, (GLXContext)ExposedData.OpenGLLinux.X11Context))
-		{
-			os::Printer::log("Render Context switch failed.");
-			return false;
-		}
-		else
-		{
-			Drawable = ExposedData.OpenGLLinux.X11Window;
-			X11Display = (Display*)ExposedData.OpenGLLinux.X11Display;
-		}
-	}
-	return true;
-}
-
-
-//! inits the open gl driver
-bool COpenGLDriver::initDriver(CIrrDeviceLinux* device)
-{
-	ExposedData.OpenGLLinux.X11Context = glXGetCurrentContext();
-	ExposedData.OpenGLLinux.X11Display = glXGetCurrentDisplay();
-	ExposedData.OpenGLLinux.X11Window = (unsigned long)Params.WindowId;
-	Drawable = glXGetCurrentDrawable();
-	X11Display = (Display*)ExposedData.OpenGLLinux.X11Display;
-
-    AuxContexts = auxCtxts;
+#if defined(_NBL_COMPILE_WITH_WINDOWS_DEVICE_)
+        HDC hdc = platform_dependent.surface;
+        HGLRC hrc = platform_dependent.context;
+        ocl_init = ocl::COpenCLHandler::getCLDeviceFromGLContext(clDevice, clProperties, hrc, hdc);
+#elif defined(_NBL_COMPILE_WITH_X11_DEVICE_)
```

Comment:
I do understand that you need to get the ACTUAL fake-EGL contexts and displays, but should w edo something about the `_NBL_COMPILE_WITH` macros?

Reply:
this is tyhe only place where theyre needed in whole driver

Reply:
I'm just wondering if we can turn this more intoa runtime check

btw Linux could be compiled both with X11 and EGL (Wayland) so this macro "elif" is probably not the healthiest solution

Reply:
this is WGL/GLX choice, not x11/EGL

Reply:
yeah but if you want OpenCL working everywhere, then you'll need to do this for all possible window systems ;)

Reply:
this code will be dead ina few days anyway
+ from what i see inside `ocl::COpenCLHandler::getCLDeviceFromGLContext`, you have to explicitely specify to ocl whether its glx or wgl or whatever else, so whats the differemce

Reply:
`ocl::COpenCLHandler` will need to be extended to work with more display systems

Reply:
leaving this comment as open

Reply:
made an issue out of it
https://github.com/Devsh-Graphics-Programming/Nabla/issues/116